abci/precommit/precommit.go

```diff
@@ -0,0 +1,96 @@
+package precommit
+
+import (
+	"cosmossdk.io/log"
+	"github.com/cosmos/cosmos-sdk/types/module"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/skip-mev/block-sdk/block"
+)
+
+// MempoolEvictionPreCommiter is a PreCommit function that evicts txs that invalid from the app-side mempool.
+type MempoolEvictionPreCommiter struct {
+	// logger
+	logger log.Logger
+
+	// app side mempool interface
+	mempl block.Mempool
+
+	// module manager
+	mm *module.Manager
+
+	// ante handler
+	anteHandler sdk.AnteHandler
+}
+
+// NewMempoolEvictionPreCommiter returns a new MempoolEvictionPreCommiter handler.
+func NewMempoolEvictionPreCommiter(
+	logger log.Logger,
+	mempl block.Mempool,
+	mm *module.Manager,
+	anteHandler sdk.AnteHandler,
+) MempoolEvictionPreCommiter {
+	return MempoolEvictionPreCommiter{
+		logger:      logger,
+		mempl:       mempl,
+		mm:          mm,
+		anteHandler: anteHandler,
+	}
+}
+
+// PreCommit returns a PreCommit handler that wraps the default PreCommit handler and evicts invalid txs from the
+// app-side mempool
+func (m *MempoolEvictionPreCommiter) PreCommit() sdk.Precommiter {
+	return func(ctx sdk.Context) {
+		defer func() {
+			if rec := recover(); rec != nil {
+				m.logger.Error(
+					"panic in precommiter",
+					"err", rec)
+			}
+		}()
```

Comment:
`panic` -> `recover` is necessary in this function because `sdk.Precommiter` does not return an error